airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py

```diff
@@ -187,34 +187,43 @@ def get_json_schema(self) -> JsonSchema:
     def _get_raw_json_schema(self) -> JsonSchema:
         if self.config.input_schema:
             return self.config.get_input_schema()  # type: ignore
-        elif self.config.schemaless:
+        if self.config.schemaless:
             return schemaless_schema
+
+        if self.config.files_to_read_for_schema_discover:
+            self.logger.info(
+                msg=(
+                    f"Only first {self.config.files_to_read_for_schema_discover} files will be used to infer schema "
+                    f"for stream {self.name} due to limitation in config."
+                )
+            )
+            files = self.list_first_n_files(self.config.files_to_read_for_schema_discover)
         else:
             files = self.list_files()
-            total_n_files = len(files)
-
-            if total_n_files == 0:
-                self.logger.warning(msg=f"No files were identified in the stream {self.name}. Setting default schema for the stream.")
-                return schemaless_schema
-
-            max_n_files_for_schema_inference = self._discovery_policy.get_max_n_files_for_schema_inference(self.get_parser())
-            if total_n_files > max_n_files_for_schema_inference:
-                # Use the most recent files for schema inference, so we pick up schema changes during discovery.
-                files = sorted(files, key=lambda x: x.last_modified, reverse=True)[:max_n_files_for_schema_inference]
-                self.logger.warn(
-                    msg=f"Refusing to infer schema for all {total_n_files} files; using {max_n_files_for_schema_inference} files."
-                )
 
-            inferred_schema = self.infer_schema(files)
+        total_n_files = len(files)
+        if total_n_files == 0:
+            self.logger.warning(msg=f"No files were identified in the stream {self.name}. Setting default schema for the stream.")
+            return schemaless_schema
 
-            if not inferred_schema:
-                raise InvalidSchemaError(
-                    FileBasedSourceError.INVALID_SCHEMA_ERROR,
-                    details=f"Empty schema. Please check that the files are valid for format {self.config.format}",
-                    stream=self.name,
-                )
+        max_n_files_for_schema_inference = self._discovery_policy.get_max_n_files_for_schema_inference(self.get_parser())
+        if total_n_files > max_n_files_for_schema_inference:
+            # Use the most recent files for schema inference, so we pick up schema changes during discovery.
+            files = sorted(files, key=lambda x: x.last_modified, reverse=True)[:max_n_files_for_schema_inference]
```

Comment:
@brianjlai when the file-based CDK was created there was no way to request the files in a specific order. But it definitely makes sense to check to see if this has changed. If it has we'll want to update the way we do `read` too so we don't have to page through everything.

Assuming the API hasn't changed, instead of just giving users a warning that schema discovery could be less accurate, I'm wondering if we should surface the option to still list & sort all files even if the user is setting a limit for schema discovery. It seems possible to me that for most customers the bulk of the time spent on `discover` is spent on reading e.g. hundreds of very large files rather than paging through the list of millions of files (obviously this will vary from customer to customer though).

I'm loathe to add too many options, but since the consequence of a bad schema is pretty bad (dropped records) it feels like we want to make sure we use the most recent records as much as possible.

Reply:
> I'm wondering if we should surface the option to still list & sort all files even if the user is setting a limit for schema discovery, in addition to your suggestion that we limit the pages of files read. It seems possible to me that for most customers the bulk of the time spent on discover is spent on reading

So if I'm understanding correctly, we still fetch every single file, but still sort them. And then once sorted, we only read X most recent files according to the config?

I think that would make sense, and in this case we don't need a new config option right? The presence of `files_to_read_for_schema_discovery` should be enough to indicate that we only do partial schema discovery of the most recent. It feels like we shouldn't even give users an option and the default behavior if they specify this new field is to first list & sort all files. Then do schema discovery.

I guess the only gap is if there way too many files to fetch before timing out since fetching files is a prerequisite to the sort, although maybe we don't need to solve this problem perfectly